airflow/upgrade/rules/customoperator_uses_metaclass.py

```diff
@@ -0,0 +1,39 @@
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+from __future__ import absolute_import
+
+from airflow.models import BaseOperator
+from airflow.models.baseoperator import BaseOperatorMeta
+from airflow.upgrade.rules.base_rule import BaseRule
+
+
+class CustomOperatorUsesMetaclassRule(BaseRule):
+
+    title = "BaseOperator uses metaclass"
+
+    description = """\
+BaseOperator class uses a BaseOperatorMeta as a metaclass. This metaclass is based on abc.ABCMeta.
+
+If your custom operator uses different metaclass then you will have to adjust it."""
+
+    def check(self):
+        if not isinstance(BaseOperator, BaseOperatorMeta):
```

Comment:
I'm afraid that `BaseOperatorMeta` is not available in Airflow 1.10.X so we cannot use it like that. 

Approaches that comes to my mind:
- create a dagbag of users DAGs and then iterate over the tasks and check if any of them is using metaclass (any of them as Python allow users using only a single metaclass)
- find all import statements in users DAGs and analyze those classes that inherit from `BaseOperator`

We can also just rise this as a warning without any check as using metaclass is rather an edge case imho. @kaxil @mik-laj what do you think?

Reply:
I do not think this is an edge case. Many users use the ABC of the metaclass. This may not be common in this project, but people who know Jave (previously used Spark) and then started writing operators may be overusing abstract classes.